Topics: add pooled mining (#1499)
bitcoinops/bitcoinops.github.io: _topics/en/pooled-mining.md

    url: /en/newsletters/2021/09/08/#braidpool-a-p2pool-alternative

## Optional.  Same format as "primary_sources" above
#see_also:
#  - title:
#    link:

## Optional.  Force the display (true) or non-display (false) of stub
## topic notice.  Default is to display if the page.content is below a
## threshold word count
#stub: false

## Required.  Use Markdown formatting.  Only one paragraph.  No links allowed.
## Should be less than 500 characters
excerpt: >
  **Pooled mining** occurs when two or more independent miners
  collaborate on finding proof of work for new blocks, with them fairly
  dividing the rewards of any blocks they find.  **Betterhash**,
  **Braidpool**, **Stratum**, and **Stratum v2** are protocols for
  coordinating pooled mining.  A **block withholding attack** is a
  well-known attack against pooled mining.

---
Hashing a block header will produce a value that can be interpreted as a
number.  Hashing many unique block headers using double-SHA256 will
produce a uniform distribution of numbers within the range 0 to
2<sup>256</sup>-1.  For example, approximately 1% of the numbers will
have values less than 1% of the range's maximum value.  For a block to

> contain enough proof of work to be valid, it's header must hash to a

Reviewer: ```suggestion
contain enough proof of work to be valid, its header must hash to a
```

> value below a _target_ value that is the dynamically set by the

Reviewer 2: ```suggestion
value below a _target_ value that is dynamically set by the
```

to profit even if she doesn't personally find a hash with the 10x higher
amount of proof of work needed for a valid block.  If Alice does find
a proof of work demonstrating that required 10x higher value, a pure PPS
scheme will still only pay her for one share (e.g. 10%).  PPS schemes
face various problems in practice, some of which we'll describe later in
this article, but PPS provides a simple framework for understanding how
miners can equitably divide the rewards from producing blocks.

Shares are a type of _weak block_.  They need to follow at least two
rules:

- A share must be either a valid block or what would have been a valid
  block if it had a header hash below the primary target.

- The coinbase transaction must follow a template provided by the pool.
  For example, all pools we know of use the coinbase transaction to pay
  either the pool operator (who later distributes rewards to pool
  members) or to pay pool members directly.

Pools may impose additional requirements on their members, either
explicitly or by using a pool protocol that does not give members the
flexibility to make certain choices.

> ## Block withholding attacks

Reviewer 2: This is fascinating; I wasn't aware of this attack. Would it be worthwhile to add a sentence or two explaining _why_ Mallory might want to do this? What does she get out of it? I'm assuming the answer is that _other miners_ might be Mallory, so this attack should increase the fraction of blocks that they find. But maybe it's worth making that explicit (assuming I'm correct)?

> Pooled mining is fundamentally vulnerable to a problem knows as a _block

Reviewer: ```suggestion
Pooled mining is fundamentally vulnerable to a problem known as a _block
```

withholding attack_.  There are two types of shares: those that have
enough work to be a valid block and those that don't.  In a pure
pay-per-share (PPS) scheme, the pool pays the same amount for both types
but the pool only earns rewards from the Bitcoin system for producing
valid blocks.  This can allow a pool member to forfeit a small amount of
income in order to deny the pool a large amount of income.

> For example, image the ratio between the primary target and secondary

Reviewer 2: ```suggestion
For example, imagine the ratio between the primary target and secondary
```
(or "suppose")

enough proof of work to be valid blocks.  On average, Mallory gets
paid for 999 blocks out of every 1,000 she generates, reducing her
income by 0.1% but costing the pool 100% of a block reward for every
1,000 shares generated by Mallory.

We are unaware of any deployed method that fully addresses this problem.
A common mitigation is for pools to pay more for shares that have enough
work to be a valid block (which implies paying less for other shares).
Some pools also restrict who can join their pools and use techniques
from statistical analysis to attempt to detect block withholders.

## Pool protocols

The earliest pool protocol was built around transmitting block header
templates in the format used by the early Bitcoin Core RPC `getwork`.
When the speed of mining equipment made that impractical, an early
attempt at creating a distributed mining protocol was specified in BIPs
[22][BIP22] and [23][BIP23], with a partial implementation in Bitcoin
Core as the `getblocktemplate` RPC.  GetBlockTemplate never saw
widespread usage as a pool protocol itself, although many centralized
pool servers have used it for the past decade in the background to obtain

> a good set of transactions to mine from their local Bitcoin Core mode.

Reviewer: ```suggestion
a good set of transactions to mine from their local Bitcoin Core node.
```

As the GetWork-based protocol became problematic, most mining pools
switched to the Stratum protocol (v1).  Stratum v1 allows pools to
only send their members a template for a block header and a coinbase
transaction, which can be less than 200 bytes.  When a member finds a
share, they can send back roughly the same information.  This very
compact information minimizes the overhead of the pool protocol.  The

> downside of the way GetBlock and Stratum v1 are typically used is that

Reviewer: GetBlockTemplate or GetWork?

```suggestion
downside of the way GetBlock and Stratum v1 are typically used is that
```

Author: GetBlock is correct.  If you received GetBlockTemplate from your pool server, you'd have a complete list of every transaction in the template, so blind mining wouldn't be a concern.  (But AFAIK no pool server has ever sent GBT to its members; that would be bandwidth intensive---GBT can be 2x the 3x the size of the candidate block itself.)

> pool members have no insight or direct control what they mine.  They

Reviewer 2: ```suggestion
pool members have no insight or direct control over what they mine.  They
```

aren't directly informed about which transactions are included in the
template block and the pool may reject any attempt they make to change

> those transactions.  They are (by necessity) informed of what their

Reviewer 2: It is also worth mentioning that the pool members can't know which transactions are _missing_, i.e. if the pool is censoring transactions?

the past to [steal][betcoindice] from website operators and lose money
accidentally [violating consensus rules][jul2015 chainsplits].  Although
BIPs [40][bip40] and [41][bip41] were reserved for Stratum v1,
documentation for them has never been provided.

[Betterhash][] was a proposed mining pool protocol that allowed pool
members to select which transactions to include in the blocks they
create.  The protocol also made it easy for pool members who found a
new block to submit it directly to the network through their Bitcoin
full node, which could reduce propagation times and improve the chance
of that block becoming a permanent part of the block chain.  Additional
improvements in Betterhash focused on security and efficiently
distributing work across multiple mining hardware devices.  Although the
proposal has not been withdrawn, it does appear that development of it
has been replaced by development on Stratum v2.

[Stratum v2][] is an entirely new protocol developed by several of the
same people who contributed to Stratum v1.  It provides several of the
same advantages as Betterhash, although sometimes using different

> mechanisms.  Like Betterhash, one of its key advantages is that it can
> allow individual pool members to chose which transactions to include in

Reviewer 2: ```suggestion
mechanisms.  Like Betterhash, one of its key advantages is that it
allows individual pool members to choose which transactions to include in
```
(unless Stratum v2 can be run in a way that _doesn't_ allow members to choose transactions?)

Reviewer 3: > ```suggestion
> mechanisms.  Like Betterhash, one of its key advantages is that it
> allows individual pool members to choose which transactions to include in
> ```
> (unless Stratum v2 can be run in a way that _doesn't_ allow members to choose transactions?)

SV2 is a modular protocol, so it could be configured and used with encryption and binary framing capabilities, but without allowing individual miners to create and work on custom block templates. Of course the best configuration is the one in which miners select txs and work on custom block templates.
To dig more into it, every possible protocol scenario can be found at stratumprotocol.org

Author: Leaving text as-is.  I wrote this in the hypothetical voice becaues, as @GitGab19 says, pools can disallow independent transaction selection (either through the protocol or though simply not paying for any shares submitted with unapproved transactions).

their blocks.  Stratum v2 is used in production as of this writing,
although some advanced features are not fully supported by any widely
deployed mining software.

[BraidPool][] is a proposed design for a fully decentralized mining
pool.  Unlike a typical centralized mining pool managed with Stratum v1

> or v2, members of a BraidPool must all chose their own transactions and

Reviewer 2: ```suggestion
or v2, members of a BraidPool must all choose their own transactions and
```

funds will be paid for any valid share, preventing the pool from

> interfering with its member's choices about what transactions to mine.

Reviewer 2: ```suggestion
interfering with its members' choices about what transactions to mine.
```

BraidPool is inspired by the former [P2Pool][] decentralized mining pool

> but attempts mitigate some of P2Pool's problems, including its

Reviewer 2: ```suggestion
but attempts to mitigate some of P2Pool's problems, including its
```

pool.  Unlike a typical centralized mining pool managed with Stratum v1
or v2, members of a BraidPool must all chose their own transactions and
funds will be paid for any valid share, preventing the pool from
interfering with each individual member's choices about what transactions to mine.
BraidPool is inspired by the former [P2Pool][] decentralized mining pool
but attempts to mitigate some of P2Pool's problems, including its
reduced ability to capture fee income due to its large typical
coinbase transactions and the frequency of P2Pool shares becoming stale.
BraidPool is under active development as of this writing.

## Pool payout schemes

Previously in this article, we've describe a single idealized payout:
a miner who creates a 1/100th share for a block template with a block
reward of 12.3456789 BTC will receive a payout of 0.1234567890 BTC.  In
practice, no pool we're aware of pays exactly that way, and they use
slightly different jargon.  The following describes general types of
payouts, although each pool's implementation of them may differ from our
descriptions and from other pools.

- **Pay per share (PPS)** pays an amount proportional to each share's

>   PoW.  For example, in a pool that changes a 0% membership fee, each

Reviewer: ```suggestion
  PoW.  For example, in a pool that charges a 0% membership fee, each
```

  share might be proportional to the block subsidy.  If the subsidy is
  3.25 BTC and a miner creates a 1/100th share, the miner receives a
  payment of 0.0325 BTC.  PPS has the advantage of being extremely
  simple.  A miner can compare all PPS pools and choose the one that

>   advertises the highest rate; the miner can then simply count how many

Reviewer: Highest rate vs membership fee language above could be confusing (or is for me?)
